### PR TITLE
Kernel: Be more strict about thread states

### PR DESCRIPTION
**Pull request by implicitfield**
As discussed in #26160, it isn't possible for a thread to go right from being blocked to being stopped (or the other way around), so this PR codifes that invariant. This PR also disallows calling `Thread::set_state()` when that would turn out to be a no-op.

### CODE REVIEW

File: Kernel/Tasks/Thread.cpp [hunk 1]
```diff
@@ -518,15 +518,24 @@ StringView Thread::state_string() const
     case Thread::State::Stopped:
         return "Stopped"sv;
     case Thread::State::Blocked: {
+        return "Blocked"sv;
+    }
```
Comment (spholz) [inline Kernel/Tasks/Thread.cpp]: These braces should be unnecessary.

File: Kernel/Tasks/Thread.cpp [hunk 2]
```diff
@@ -229,17 +221,10 @@ void Thread::block(Kernel::Mutex& lock, SpinlockLocker<Spinlock<LockRank::None>>
...
-        break;
-    case Thread::State::Running:
-        VERIFY(m_blocker == nullptr);
-        break;
-    default:
-        dbgln("Error: Attempting to block with invalid thread state - {}", state_string());
-        VERIFY_NOT_REACHED();
-    }
+    if (state() != Thread::State::Running)
+        PANIC("Error: Attempting to block with invalid thread state - {}", state_string());
```
Comment (spholz) [inline Kernel/Tasks/Thread.cpp]: I would remove the "Error: " prefix.

File: Kernel/Tasks/Thread.cpp [hunk 3]
```diff
@@ -132,16 +132,8 @@ Thread::BlockResult Thread::block_impl(BlockTimeout const& timeout, Blocker& blo
...
-    case Thread::State::Stopped:
-        // It's possible that we were requested to be stopped!
-        break;
-    case Thread::State::Running:
-        VERIFY(m_blocker == nullptr);
-        break;
-    default:
-        VERIFY_NOT_REACHED();
-    }
+    VERIFY(state() == Thread::State::Running);
```
Comment (spholz) [inline Kernel/Tasks/Thread.cpp]: It might be nice to also PANIC() with a better message here, like below.

File: Kernel/Tasks/Thread.cpp [hunk 4]
```diff
@@ -898,21 +883,13 @@ static ErrorOr<void> copy_value_on_user_stack(FlatPtr& stack, T const& data)
...
-        } else {
-            // Was unblocked while stopped
-            set_state(Thread::State::Runnable);
-        }
-    } else {
-        set_state(m_stop_state, 0);
-    }
+    VERIFY(is_stopped());
+
+    if (m_stop_state != Thread::State::Running && m_stop_state != Thread::State::Runnable)
```
Comment (spholz) [inline Kernel/Tasks/Thread.cpp]: Would it be nicer to verify this while setting `m_stop_state`?

File: Kernel/Tasks/Thread.cpp [hunk 5]
```diff
@@ -229,17 +221,10 @@ void Thread::block(Kernel::Mutex& lock, SpinlockLocker<Spinlock<LockRank::None>>
     SpinlockLocker scheduler_lock(g_scheduler_lock);
     SpinlockLocker block_lock(m_block_lock);
 
-    switch (state()) {
-    case Thread::State::Stopped:
-        // It's possible that we were requested to be stopped!
```
Comment (spholz) [inline Kernel/Tasks/Thread.cpp:234]: I wonder why we had a special case for `Stopped` here. Are you sure this can't happen somehow?

**Review by spholz — COMMENTED** [review]
Thanks! I just hope that we don't have an edge case that hits these VERIFYs somehow. Though these changes would let us now that is the case 😅.

File: Kernel/Tasks/Thread.cpp (hunk 5 above)
Comment (implicitfield) [reply Kernel/Tasks/Thread.cpp:234]: No idea really why this was here. I can't see this happening in a non-SMP context really (since `Thread::block{_impl}` both take the scheduler lock as `Thread::set_state` does). I'm not sure if this could happen with SMP though (but if it can, then I feel like we should prevent this from happening on a higher level).

Comment (implicitfield) [reply Kernel/Tasks/Thread.cpp:234]: Actually, I don't think this should happen with SMP either, since it looks like a thread can only go into the `Stopped` state of its own accord since signals are only handled from within the thread itself, so there shouldn't be any parallellism-induced issues either.

**Comment by implicitfield** [comment]
The following also seems to be out of date, since blockers haven't dispatched signals since 7d894096 (and the reason why the comment talks about `Thread::block` and not `Thread::block_impl` seems to be that it predates f469fb47b as well):
https://github.com/SerenityOS/serenity/blob/6dc2aa998a40b6a3a9955d0f789aa45efe85114e/Kernel/Tasks/ThreadBlockers.cpp#L714-L722

File: Kernel/Tasks/Thread.cpp [hunk 6]
```diff
@@ -1264,6 +1240,9 @@ void Thread::set_state(State new_state, u8 stop_signal)
     } else if (m_state == Thread::State::Stopped) {
         // We don't want to restore to Running state, only Runnable!
         m_stop_state = previous_state != Thread::State::Running ? previous_state : Thread::State::Runnable;
```
Comment (spholz) [inline Kernel/Tasks/Thread.cpp]: Shouldn't this now always be `m_stop_state = Runnable` and therefore make `m_stop_state` unnecessary?

**Comment by spholz** [comment]
Sorry for the delay, I currently don't have much time for reviewing this.
I want to do a more in-depth review in ~2 weeks.

File: Kernel/Tasks/Thread.cpp (hunk 6 above)
Comment (implicitfield) [reply Kernel/Tasks/Thread.cpp]: Yeah, that does seem to be the case, since we're basically asserting that here now. (I hope this really is this straightforward, since this makes `Thread::resume_from_stopped()` so much nicer).

File: Kernel/Tasks/Thread.cpp (hunk 5 above)
Comment (spholz) [reply Kernel/Tasks/Thread.cpp:234]: So I think this should be safe as long as all threads stop themselves (which they do)
and immediately yield without calling `block{,_impl}` anywhere.
Most places seem to yield after `set_state(Stopped)`. There is one exception though: https://github.com/SerenityOS/serenity/blob/3735606ddc2cdb8e2d3783ba74aacfe6a3a469dd/Kernel/Tasks/Scheduler.cpp#L306
`dispatch_one_pending_signal()` returns a `DispatchSignalResult`, which can be `Yield` if the thread should yield. I think ignoring this result is an oversight. We should maybe even make that enum `[[nodiscard]]`.

I think this call to `dispatch_one_pending_signal` at the end of `context_switch` could actually cause a SIGSTOP to be delivered and ignored. The code after that context switch could then theoretically block while `Stopped`.

Comment (implicitfield) [reply Kernel/Tasks/Thread.cpp:234]: We're also ignoring `DispatchSignalResult` here:
https://github.com/SerenityOS/serenity/blob/2ccc394e65a6939d8a63b9e2f064f33217599a2d/Kernel/Tasks/ThreadBlockers.cpp#L698 and here:
https://github.com/SerenityOS/serenity/blob/2ccc394e65a6939d8a63b9e2f064f33217599a2d/Kernel/Tasks/Thread.cpp#L1223

The `ThreadBlockers.cpp` case might not be too bad, since it always just sends a SIGCHLD, but even that seems like it could require a yield if the process is being traced.

Not sure what to do with the other two though, since it seems we can't really just yield while the scheduler lock is held. For the scheduler case, I guess we could just disable interrupts while holding the thread lock and then release the scheduler lock, if that sounds reasonable.

The `Thread::set_state` case is kind of annoying, since it's called from so many places (including `Scheduler::context_switch`). The same thing could be applied here, but I'm wondering if it's completely right to yield in all these contexts. Though, does `Thread::set_state` actually need to dispatch signals when context switches also do that? (the considerations might have been different when c3d2316 was authored, since context switches didn't do that at the time). 

Edit: Though those two extra cases specifically happen when we're not stopped, so those are just about missed yields in general I guess.

Comment (spholz) [reply Kernel/Tasks/Thread.cpp:234]: I feel like we are dispatching signals in too many places.
Signals should only be dispatched in switches between user and kernel mode I think. So for context switches, syscalls, and exceptions.

Doing this in `set_state()` definitely seems weird.

Comment (implicitfield) [reply Kernel/Tasks/Thread.cpp:234]: I've removed a bunch of (fairly likely) unnecessary signal dispatching in the latest push. The scheduler should now also yield if a dispatched signal requires it to, but this is quite hard to test (this seems to work, but hopefully this doesn't have any odd edge cases).

File: Tests/Kernel/TestWait.cpp [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+#include <LibTest/TestCase.h>
+#include <errno.h>
+#include <signal.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/wait.h>
+
+static bool s_received_sigchld = false;
```
Comment (spholz) [inline Tests/Kernel/TestWait.cpp]: I think you usually need `volatile` variables if you modify them in the signal handler.
But since you only read it once, compiler optimizations shouldn't cause any problems I think?

File: Kernel/Tasks/Scheduler.cpp [hunk 8]
```diff
@@ -303,7 +305,12 @@ void Scheduler::context_switch(Thread* thread)
 
     {
         SpinlockLocker lock(thread->get_lock());
-        thread->dispatch_one_pending_signal();
+        InterruptDisabler disable;
+        if (thread->dispatch_one_pending_signal() == DispatchSignalResult::Yield) {
+            g_scheduler_lock.unlock(InterruptsState::Disabled);
+            yield();
+            g_scheduler_lock.lock();
+        }
```
Comment (spholz) [inline Kernel/Tasks/Scheduler.cpp]: I don't really like this code.
We should either reschedule outside of `pick_next()` if we notice that we should yield or preferably we should not dispatch signals here either.

I think that we only ever need to dispatch signals before returning to user mode. That should also work for programs that never issue syscalls, since timer interrupts exist.
I'm not really sure if that's enough though. But if it is, signal dispatching would be a lot easier. (No clue what other Unixes do)

File: Kernel/Tasks/Thread.cpp (hunk 5 above)
Comment (spholz) [reply Kernel/Tasks/Thread.cpp:234]: nit: Your commit messages say signals are only dispatched during context switches. But that's not always true. They could for example also be dispatched after returning from a syscall handler.

File: Kernel/Tasks/Scheduler.cpp (hunk 8 above)
Comment (spholz) [reply Kernel/Tasks/Scheduler.cpp]: Though doing this only before returning to user mode means that signals can have some latency. This might not be nice for urgent signals like SIGSTOP or SIGKILL. They already have some latency, but with that they could have even more latency in some cases, not sure if that matters though.

File: Kernel/Tasks/Thread.cpp [hunk 9]
```diff
@@ -1254,8 +1226,9 @@ void Thread::set_state(State new_state, u8 stop_signal)
         Scheduler::enqueue_runnable_thread(*this);
         Processor::smp_wake_n_idle_processors(1);
     } else if (m_state == Thread::State::Stopped) {
-        // We don't want to restore to Running state, only Runnable!
-        m_stop_state = previous_state != Thread::State::Running ? previous_state : Thread::State::Runnable;
+        if (previous_state != Thread::State::Running && previous_state != Thread::State::Runnable)
```
Comment (implicitfield) [inline Kernel/Tasks/Thread.cpp]: I actually can't think of a way for this to be `Runnable` now. I kind of recall seeing that happen for some vaguely understandable reason, but I can't remember the details and couldn't get that to happen anymore. (so this now only accepts `Running`).

File: Tests/Kernel/TestWait.cpp (hunk 7 above)
Comment (implicitfield) [reply Tests/Kernel/TestWait.cpp]: This ended up working in practice anyway, but yes it's probably a good idea / better stylistically to make this `volatile`.

File: Kernel/Tasks/Scheduler.cpp (hunk 8 above)
Comment (implicitfield) [reply Kernel/Tasks/Scheduler.cpp]: I kind of prefer dispatching signals during context switches, it just feels kind of like a hack to rely on timer IRQs for this. From a quick look, it seems like Linux handles signals by basically marking a thread as runnable as soon as it has a signal to respond to (so before it's actually scheduled), and then the scheduler handles it like usual. It's conceptually kind of the same as what we're doing I guess, but our signal dispatches are more explicit.

File: Kernel/Tasks/Thread.cpp (hunk 5 above)
Comment (implicitfield) [reply Kernel/Tasks/Thread.cpp:234]: I was too fixated on context switches at this point I see (I didn't really mean that that's the only way these'd get dispatched, but I see the old commit messages did pretty heavily imply that).

File: Kernel/Tasks/Scheduler.cpp (hunk 8 above)
Comment (spholz) [reply Kernel/Tasks/Scheduler.cpp]: I don't mean that we should only do this in timer IRQs. I mean that we should check for signals whenever we return to user mode.

This seems to be in line with what the [signal(7) linux man page](https://www.man7.org/linux/man-pages/man7/signal.7.html) says:
> Execution of signal handlers
>    Whenever there is a transition from kernel-mode to user-mode
>    execution (e.g., on return from a system call or scheduling of a
>    thread onto the CPU), the kernel checks whether there is a pending
>    unblocked signal for which the process has established a signal
>    handler.

Looking at OpenBSD code, `postsig()` is only called by `userret()`, which is called when returning from syscalls and exceptions that are from userspace. It's also called when returning from ASTs. ASTs are described like this by the amd64 code:
```
 *	AST handler.  This is called from assembly language stubs when
 *	returning to userspace after a syscall or interrupt.
```

So unless I'm reading this code wrong (or I'm missing something, which could totally be the case), they only dispatch signals before returning to userspace.
It could also be that they handle urgent signals somewhere else, not sure.

Comment (implicitfield) [reply Kernel/Tasks/Scheduler.cpp]: I guess that could work as well, just the problem with that is is that we don't always dispatch signals when returning to user mode (outside of the scheduler that is, otherwise this only happens in `syscall_handler`). So without this, a process that doesn't make syscalls never receives any signals. We'd basically need to dispatch signals in the ISR return to make this work without the scheduler doing this, but that'd require (trivial, but still) arch-specific changes (unless we share `exit_trap()` between all architectures).

Comment (spholz) [reply Kernel/Tasks/Scheduler.cpp]: Yeah I meant we should change the code in `exit_trap()` or similar to dispatch signals and stop dispatching signals in other places.
Or do you prefer you current approach? I think only dispatching signals in one place seems nice. Though if we go with that approach, we should probably test it thoroughly. So test that EINTR still works, signals are dispatched even if no syscalls are issued, etc.

But I also don't want to sidetrack you too much from the WaitQueue rewrite :sweat_smile:.

> unless we share `exit_trap()` between all architectures

Currently, the only difference between the x86-64 vs AArch64/RISC-V versions is that they always set `m_in_irq` to 0 and don't have `prev_irq_level` in their `TrapFrame`s. I never looked into this, but from the name of the variable I guess it's supposed to be used by nested interrupts?
If we can unify these versions of `exit_trap()` that would be nice, but you don't have to do it in this PR.

Comment (implicitfield) [reply Kernel/Tasks/Scheduler.cpp]: Just dispatching signals in one place does sound quite nice actually (I figured we'd still need to do that in both syscall and ISR exit points, but apparently `syscall_handler` goes through `exit_trap()` even on x86_64, so that shouldn't actually be necessary).

In practice, this does seem quite difficult to do. We don't have the scheduler lock outside of `pick_next()`, but I guess we could just take the lock at a higher level (since signal dispatching requires that we do have the lock). A more significant issue is that we're dispatching urgent signals in a lot of places, and replacing those with regular signal dispatches probably isn't right, since `send_urgent_signal_to_self()` is basically `[[noreturn]]` if the thread needs to terminate as a result.

As to unifying `exit_trap()`, I think that might actually be doable. I just ran through all the tests (and shut down the system afterwards) on x86_64 with `VERIFY(trap.prev_irq_level == 0)` in `exit_trap()`, and nothing ever tripped it. Since AArch64 and RISC-V work also fine without that too, then I guess it really isn't necessary?

Comment (spholz) [reply Kernel/Tasks/Scheduler.cpp]: > In practice, this does seem quite difficult to do. We don't have the scheduler lock outside of `pick_next()`, but I guess we could just take the lock at a higher level (since signal dispatching requires that we do have the lock).

Can't we just keep using `check_dispatch_pending_signal()`? That function doesn't need any locks.

> A more significant issue is that we're dispatching urgent signals in a lot of places, and replacing those with regular signal dispatches probably isn't right, since `send_urgent_signal_to_self()` is basically `[[noreturn]]` if the thread needs to terminate as a result.

I would keep using `send_urgent_signal_to_self()` to send signals caused by hardware exceptions. (That also seems to be what OpenBSD does in `trapsignal()`. This seems to be the only other function next to `postsig()` that calls `sendsig()`)
I'm a bit skeptical of `Process::tracer_trap()` and this line in `do_exec()` using `send_urgent_signal_to_self()` though:
https://github.com/SerenityOS/serenity/blob/3735606ddc2cdb8e2d3783ba74aacfe6a3a469dd/Kernel/Tasks/Process.cpp#L1111
https://github.com/SerenityOS/serenity/blob/3735606ddc2cdb8e2d3783ba74aacfe6a3a469dd/Kernel/Syscalls/execve.cpp#L614

But again, sorry if I'm sidetracking you too much. We could keep the current `ShouldYield` approach if you don't want to work on this.

> As to unifying `exit_trap()`, I think that might actually be doable. I just ran through all the tests (and shut down the system afterwards) on x86_64 with `VERIFY(trap.prev_irq_level == 0)` in `exit_trap()`, and nothing ever tripped it. Since AArch64 and RISC-V work also fine without that too, then I guess it really isn't necessary?

Yes I think that variable would only be necessary if nested interrupts would be possible. So I guess at some point we enabled interrupts during interrupt handling? `m_in_irq` can also simply a bool.

Comment (implicitfield) [reply Kernel/Tasks/Scheduler.cpp]: > Can't we just keep using `check_dispatch_pending_signal()`? That function doesn't need any locks.

It doesn't directly, but it can call `dispatch_one_pending_signal()` which does `VERIFY(g_scheduler_lock.is_locked_by_current_processor())`.

---

I guess it might then be fine to use `send_urgent_signal_to_self()` in most cases, but I guess there's probably no need to use that for `SIGTRAP` or `SIGSTOP`(or anything that doesn't immediately terminate the process really).

> But again, sorry if I'm sidetracking you too much. We could keep the current ShouldYield approach if you don't want to work on this.

It's no problem really - though It seems like things don't work quite right with the approach I'm currently trying (just using a loop in `Scheduler::invoke_async()` since that's what `exit_trap()` ends up calling through `check_invoke_scheduler()`), so maybe it's best to use the `ShouldYield` approach for now.

> Yes I think that variable would only be necessary if nested interrupts would be possible. So I guess at some point we enabled interrupts during interrupt handling? `m_in_irq` can also simply a bool.

That's probably what that was for then. There's also this comment in in `Kernel/Arch/Processor.h` (above `m_in_irq`):
```
// FIXME: On aarch64, once there is code in place to differentiate IRQs from synchronous exceptions (syscalls),
//        this member should be incremented. Also this member shouldn't be a FlatPtr.
```
But the essence of that has been there since 2022, so I'm not sure if that's really relevant anymore.
(I already have local changes that unify TrapFrame & exit_trap() across all architectures, so if this doesn't block that, then I can send another PR for that independently of anything else).

Comment (spholz) [reply Kernel/Tasks/Scheduler.cpp]: > It doesn't directly, but it can call `dispatch_one_pending_signal()` which does `VERIFY(g_scheduler_lock.is_locked_by_current_processor())`.

The `g_scheduler_lock` is locked by it here:
https://github.com/SerenityOS/serenity/blob/3735606ddc2cdb8e2d3783ba74aacfe6a3a469dd/Kernel/Tasks/Thread.cpp#L641

> I guess it might then be fine to use `send_urgent_signal_to_self()` in most cases, but I guess there's probably no need to use that for `SIGTRAP` or `SIGSTOP`(or anything that doesn't immediately terminate the process really).

I would keep using `send_urgent_signal_to_self()` for `SIGTRAP`s that are generate by hardware breakpoint exceptions but maybe not for the other one I noted.

> It's no problem really - though It seems like things don't work quite right with the approach I'm currently trying (just using a loop in `Scheduler::invoke_async()` since that's what `exit_trap()` ends up calling through `check_invoke_scheduler()`), so maybe it's best to use the `ShouldYield` approach for now.

I think I would do it after the call to `send_urgent_signal_to_self()` in `exit_trap()` if the previous mode is `ExecutionMode::User`?

> That's probably what that was for then. There's also this comment in in `Kernel/Arch/Processor.h` (above `m_in_irq`):
> 
> ```
> // FIXME: On aarch64, once there is code in place to differentiate IRQs from synchronous exceptions (syscalls),
> //        this member should be incremented. Also this member shouldn't be a FlatPtr.
> ```
> 
> But the essence of that has been there since 2022, so I'm not sure if that's really relevant anymore. (I already have local changes that unify TrapFrame & exit_trap() across all architectures, so if this doesn't block that, then I can send another PR for that independently of anything else).

That comment is actually slightly wrong. It has to be a `FlatPtr` since x86-64 uses `read_gs_ptr` to read it.

Comment (implicitfield) [reply Kernel/Tasks/Scheduler.cpp]: > The g_scheduler_lock is locked by it here:

Ah, sure, I somehow managed to miss that.

> I would keep using send_urgent_signal_to_self() for SIGTRAPs that are generate by hardware breakpoint exceptions but maybe not for the other one I noted.

I take it that's for the benefit of kernel threads then? Otherwise, wouldn't dispatching signals in `exit_trap()` have the same effect for user-mode threads?

> I think I would do it after the call to send_urgent_signal_to_self() in exit_trap() if the previous mode is ExecutionMode::User?

Sorry, I'm not seeing where in `exit_trap()` we're calling `send_urgent_signal_to_self()`. Maybe signal dispatching could be done right after this line, with a check for `!current_trap` since that seems to signal a switch to user-mode:
https://github.com/SerenityOS/serenity/blob/2ccc394e65a6939d8a63b9e2f064f33217599a2d/Kernel/Arch/x86_64/Processor.cpp#L832

> That comment is actually slightly wrong. It has to be a FlatPtr since x86-64 uses read_gs_ptr to read it.

I recalled something similar, but I guess the part about incrementing that variable also kind of unnecessary if it's really just for nested IRQs that don't actually happen.

Comment (spholz) [reply Kernel/Tasks/Scheduler.cpp]: > I take it that's for the benefit of kernel threads then? Otherwise, wouldn't dispatching signals in `exit_trap()` have the same effect for user-mode threads?

I just thought that we should treat breakpoint exceptions the same as illegal instruction exceptions (as both cause a crash by default), but I guess since those exceptions don't use `handle_crash()`, we could also dispatch them in `exit_trap()`.

> Sorry, I'm not seeing where in `exit_trap()` we're calling `send_urgent_signal_to_self()`.

Oh oops I meant `check_invoke_scheduler()`.

> Maybe signal dispatching could be done right after this line, with a check for `!current_trap` since that seems to signal a switch to user-mode:

The first (highest level) trap can also be caused by kernel mode. I think we should check that the trap that we will return from was from user mode.

Comment (implicitfield) [reply Kernel/Tasks/Scheduler.cpp]: For some reason this still doesn't seem to work. Just dispatching signals in `check_invoke_scheduler()` with `check_dispatch_pending_signal()` breaks the Shell (it's kind of unresponsive, and PS1 is red for some reason), and probably other things too.
Edit: Shell works slightly better if we dispatch signals after invoking the scheduler, but the issue with PS1 remains.

Comment (spholz) [reply Kernel/Tasks/Scheduler.cpp]: Sure, then let's keep the current approach for now.

File: Kernel/Tasks/ThreadBlockers.cpp [hunk 10]
```diff
@@ -680,22 +680,14 @@ void Thread::WaitBlocker::will_unblock_immediately_without_blocking(UnblockImmed
...
 
     if (try_unblock)
         Process::current().wait_blocker_set().try_unblock(*this);
-
-    // If we were interrupted by SIGCHLD (which gets special handling
-    // here) we're not going to return with EINTR. But we're going to
-    // deliver SIGCHLD (only) here.
-    auto* current_thread = Thread::current();
-    if (got_sigchld && current_thread->state() != State::Stopped)
-        current_thread->try_dispatch_one_pending_signal(SIGCHLD);
```
Comment (spholz) [inline Kernel/Tasks/ThreadBlockers.cpp:698]: `try_dispatch_one_pending_signal` is now unused with this PR.

**Review by spholz — APPROVED** [review]
Thanks for pushing this through!
